From 21eab7db31138b745c4acc1d59c9f6340f746b0a Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Sat, 19 Jul 2014 04:25:08 -0400
Subject: [pci] msi: Return msix_capability_init() failure if populate_msi_sysfs() fails

Message-id: <20140719042508.19438.92097.stgit@gir.stowe>
Patchwork-id: 85985
O-Subject: [RHEL7.1 PATCH 23/81] PCI/MSI: Return msix_capability_init() failure if populate_msi_sysfs() fails
Bugzilla: 1110895
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110895
Upstream Status: 2adc7907bac2c72535894732c4b41f9210f9e577

commit 2adc7907bac2c72535894732c4b41f9210f9e577
Author: Alexander Gordeev <agordeev@redhat.com>
Date:   Mon Dec 16 09:34:56 2013 +0100

    PCI/MSI: Return msix_capability_init() failure if populate_msi_sysfs() fails

    If populate_msi_sysfs() function failed msix_capability_init() must return
    the error code, but it returns the success instead.  This update fixes the
    described misbehaviour.

    Signed-off-by: Alexander Gordeev <agordeev@redhat.com>
    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>
    Reviewed-by: Tejun Heo <tj@kernel.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/pci/msi.c b/drivers/pci/msi.c
index f2ba924..36d878c 100644
--- a/drivers/pci/msi.c
+++ b/drivers/pci/msi.c
@@ -780,7 +780,7 @@ static int msix_capability_init(struct pci_dev *dev,
 
  ret = arch_setup_msi_irqs(dev, nvec, PCI_CAP_ID_MSIX);
  if (ret)
-  goto error;
+  goto out_avail;
 
  /*
   * Some devices require MSI-X to be enabled before we can touch the
@@ -793,10 +793,8 @@ static int msix_capability_init(struct pci_dev *dev,
  msix_program_entries(dev, entries);
 
  ret = populate_msi_sysfs(dev);
- if (ret) {
-  ret = 0;
-  goto error;
- }
+ if (ret)
+  goto out_free;
 
  /* Set MSI-X enabled bits and unmask the function */
  pci_intx_for_msi(dev, 0);
@@ -807,7 +805,7 @@ static int msix_capability_init(struct pci_dev *dev,
 
  return 0;
 
-error:
+out_avail:
  if (ret < 0) {
   /*
    * If we had some success, report the number of irqs
@@ -824,6 +822,7 @@ error:
    ret = avail;
  }
 
+out_free:
  free_msi_irqs(dev);
 
  return ret;
-- 
1.7.1